### ISSUE TITLE
oversized fit?

### ISSUE THREAD

**Issue by Lplantard** [fmi-faim/napari-psf-analysis#4]
I compared the results from napari-psf-analysis with other software measuring PSF, with the same data.   napari-psf-analysis was systematically giving larger values in X, Y, and Z. Could it be a problem of the fitting?

**Comment by tibuch** [comment]
Thank you for validating the results. I looked into this and simulated different 3D Gaussian blobs and fitted them with the napari-psf-analysis plugin and compared the results to the PSF analysis macro of @lgelman. 

## Testing with simulated data
### psf_11774_9419_7064_rot0_bg0.tif
![image](https://user-images.githubusercontent.com/5450869/171593618-5faccedc-193f-4a7e-961a-ca0c25439d5c.png)

This is a PSF simulation with a 3D Gaussian with sigmas (z, y, x) = (5, 4, 3). The Gaussian is axis aligned and background is set to 0.
The expected full width half maximum (FWHM) values for a voxel size of 1000x1000x1000nm would be:
* z: 11774nm
* y: 9419nm
* x: 7064nm
These values are obtained by compute FWHM from the simulation sigmas with the formula: `2 * np.sqrt(2 * np.log(2)) * sigma`
The result must be rescaled by a factor 1000 to adjust for the voxel size.

#### napari-psf-analysis results 
FWHM (z, y, x) = (11774, 9419, 7064); Window size is 30'0000nm.

#### PSF analysis macro results
FWHM (z, y, x) = (11803, 9637, 7228); Window size is 30'0000nm.

For this test the napari-psf-analysis results are correct i.e. as expected looking at the simulation. However, the PSF analysis macro returns systematically larger numbers. 

### psf_11774_9419_7064_rot45_bg0.tif
![image](https://user-images.githubusercontent.com/5450869/171596075-8d6e7d42-4ada-418d-bd6f-f6a2bbcad0b7.png)
3D Gaussian with sigmas (z, y, x) = (5, 4, 3), however the Gaussian is rotated in the XY-plane by 45 degrees.
As far as I know all the tools we use for PSF analysis compute the FWHM axis aligned, which results in a handle for the resolution along Z, Y, X axis. 
For this we should expect the same FWHM in Z and same FWHM for X and Y because of the 45 degree rotation.

#### napari-psf-analysis results
FWHM (z, y, x) = (11759, 8146, 8146); Window size is 30'0000nm.

#### PSF analysis macro results
FWHM (z, y, x) = (11848, 8405, 8405); Window size is 30'0000nm.

For this test the napari-psf-analysis result in Z is slightly too low. For all quantities the PSF analysis macro results are higher than the ones of the napari-psf-analysis plugin. Both version create equal results for FWHM of X and Y. 

### Increase background to 10
* psf_11774_9419_7064_rot0_bg10.tif
* psf_11774_9419_7064_rot45_bg10.tif
Are the same as above but with background set to 10 instead of 0. The numbers obtained by both computations are the same as shown above. Which means that background has no influence on the FWHM computation i.e. both versions take background into account.

## Why are the results different?
The formula I use to compute FWHM (see above) is based on https://en.wikipedia.org/wiki/Gaussian_function#Properties.
I looked through the macro code and identified lines 210, 280 and 281 to be doing the FWHM computation for Z, X and Y.

```
HM=(Ymax-Ymin)/2; k=2*d*d*log((HM-a)/(b-a)); FWHM=2*zVoxel*sqrt(k);
k=2*d*d*log((HM-a)/(b-a)); FWHMx=2*xyVoxel*sqrt(k);
ky=2*dy*dy*log((HM-ay)/(by-ay)); FWHMy=2*xyVoxel*sqrt(ky);
```
This formulation looks almost equal to the one I have used. However, I use `log(2)` and in the script `log((HM-a)/(b-a))` is used. Unfortunately I don't know where this formulation originates from.

_But if I replace `log((HM-a)/(b-a))` by `log(2)` the results obtained with the macro are equal to the napari-psf-analysis results._

## Conclusion
Both methods produce a correct fit (for this test data). However, they have different formulas to compute the FWHM. 
Could you point me to some other tools? Then I can compare my results to them as well.
It would also be interesting to learn about the origin of `log((HM-a)/(b-a))`.

If you can share one of the example images I could run the same tests.

The simulated PSF tiffs:
[simulated_psfs.zip](https://github.com/fmi-faim/napari-psf-analysis/files/8822545/simulated_psfs.zip)

**Comment by tibuch** [comment]
I did more testing with a real PSF:
[W1_Merian_40x_2_confGFP.zip](https://github.com/fmi-faim/napari-psf-analysis/files/8822895/W1_Merian_40x_2_confGFP.zip)

With this image I can reproduce the issue described by @Lplantard i.e. the numbers by the napari plugin are considerably larger than the ones from the macro by @lgelman. Also if I change the macro to use `log(2)` the macro is still producing smaller numbers compared to the napari plugin. 

I will continue debugging.

**Comment by tibuch** [comment]
Third time's the charm :shipit: 

Looked deeper into the macro and compared it to the napari-plugin. The macro fits a 1D Gaussian for every dimension to the pixel array crossing through the estimated PSF center pixel. The napari-plugin on the other hand fits a 3D Gaussian to a small sub-volume around the PSF. 

If the PSF (simulated or not) is exactly aligned with the pixel grid both methods obtain the same fit (up to the `log(2)`) question. Which is what I confirmed above, because the simulated Gaussians had there centers exactly at pixel-coordinates (50, 50, 50). However, if the PSF is at a sub-pixel location e.g. (50.5, 50.5, 50.5) the macro will fit a Gaussian to the intensity values of the 1D pixel arrays crossing through (50, 50, 50) whereas the napari-plugin will fit the 3D Gaussian to the whole sub-volume and obtain a sub-pixel accurate fit.

This also explains the much smaller errors in the macro fits e.g. here:
![image](https://user-images.githubusercontent.com/5450869/171613651-57cfe524-8dc7-48dd-b5b5-a2d686cb1c02.png)

Compared to the fitting errors of the napari-plugin:
![image](https://user-images.githubusercontent.com/5450869/171613798-39edf980-5b8c-46f2-b918-040bb2e623c7.png)

At first glance the macro result looks like a better fit, but I would argue that the 3D sub-pixel accurate fit is a better representation of reality given that PSF centers can not be guaranteed to be integer. 
This is also confirmed by running the analysis with a simulated bead.

### psf_4710_3061_1884_rot0_bg10_subpix.tif
This simulated PSF is centered at (50.5, 50.5, 50.5) with sigmas (z, y, x) = (2, 1.3, 0.8) which would result in expected FWHM values of:
* z: 4710
* y: 3061
* x: 1884

#### napari-psf-analysis results
FWHM (z, y, x) = (4710, 3061, 1884); Window size is 20'0000nm.

#### PSF analysis macro results
FWHM (z, y, x) = (4739, 2798, 1894); Window size is 20'0000nm.

## Conclusion
In my opinion the napari-psf-analysis plugin produces more accurate results. However, the question of `log(2)` vs. `log((HM-a)/(b-a))` remains.

**Comment by tibuch** [comment]
Fixed with #5.
